enhancements/namespace-auto-pruning.md

```diff
@@ -0,0 +1,170 @@
+
+### Goals
+
+* User is able to provide a time based policy for auto-pruning tags within a namespace
+* User is able to provide a tag number based policy for auto-pruning tags within a namespace
+* User is able to specify whether tags go into the time machine or are immediately deleted
+* The correct tags are deleted when the policy is applied
+* Pruned tags are tracked via usage logs
+
+### Non-Goals
+
+* Allowing users to specify multiple policies per namespace
+* Applying policies based on repo or tag name pattern matching
+* Allowing policy configurations at the repository level
+
+## Proposal
+
+The details in this proposal is only for namespace auto-pruning policies but designed in a way to support repository auto-pruning policies in the future. This proposal is structured in three parts: how policies are managed, how the policies are structured in the DB, and how the policies will be executed via the auto-prune worker.
+
+### Managing Policies
+
+A user is able to configure auto-pruning within the namespace settings. To start, there will be two auto-pruning options:
+* Prune by number of tags, provided the number of tags
+* Prune tags older than date, provided the number of days 
+
+For this implementation users will only be able to configure one policy per namespace. To interface with this new UI there will be a new set of CRUD endpoints for managing auto-pruning policies. The endpoints returning lists will be paginated.
+
+**Create policy**
+```
+POST /api/v1/organization/{organization}/autoprunepolicy
+By date:
+{
+    pruneBy: "creation_date",
+    olderThan: "2w"  
+}
+By number of tags:
+{
+    pruneBy: "number_of_tags",
+    numTags: 100  
+}
+```
+
+**Update policy**
+```
+PATCH /api/v1/organization/{organization}/autoprunepolicy/{policyid}
+{
+    fieldToBeUpdated: value,
+}
+```
+
+**List policies**
+```
+GET /api/v1/organization/{organization}/autoprunepolicy
+```
+
+**Get policy**
+```
+GET /api/v1/organization/{organization}/autoprunepolicy/{policyid}
+```
+
+**Delete policy**
+```
+DELETE /api/v1/organization/{organization}/autoprunepolicy/{policyid}
+```
+
+### Database structure
+Two tables will be required.
+
+**namespaceautoprunepolicy**
+
+This table holds the policy configuration for a single namespace. For this proposal there will be only one entry per namespace but there will be support for multiple rows per namespace_id. The config field will hold the policy details such as `{pruneBy: "creation_date", olderThan: "2w"}` or `{pruneBy: "number_of_tags",  numTags: 100}`.
+
+| Field | Datatype | Attributes | Description|
+| --- | ----------- | ----------- | ----------- |
+| uuid | character varying(255) | Unique, not null, indexed | Unique identifier for this policy |
+| namespace_id | integer | FK, not null | Namespace that the policy falls under |
+| config | text | JSON, not null | Policy configuration |
+
+**autoprunetask**
```

Comment:
maybe `autoprunetaskstatus` like we have `manifestsecuritystatus`

Reply:
renamed